src/telegram/process_update_task.rs

```diff
@@ -119,11 +165,224 @@ impl ProcessUpdateTask {
+        Ok(())
+    }
+
+    // What we do if users write /cancel in any state
+    async fn cancel(
+        db_controller: &DbController,
+        message: &Message,
+        api: &AsyncApi,
+    ) -> Result<(), BotError> {
+        let (chat_id, user_id, username) = Self::get_info_from_message(message);
+
+        let text = format!("Hi, {}!\n Your operation was canceled", username);
+        Self::send_message(message, &text, api).await?;
+
+        db_controller
+            .modify_before_state(&chat_id, user_id, ClientState::Initial)
+            .await?;
+
+        db_controller
+            .modify_state(&chat_id, user_id, ClientState::Initial)
+            .await?;
+
+        Ok(())
+    }
+
+    async fn find_city(
+        db_controller: &DbController,
+        username: &str,
         message: &Message,
-        text: &str,
         api: &AsyncApi,
     ) -> Result<(), BotError> {
+        let pattern = message.text.as_ref().unwrap();
+
+        let vec = db_controller.get_city_by_pattern(pattern).await?;
+
+        if vec.is_empty() || vec.len() > 30 {
+            let text = format!(
+                "Hi, {}! Your city {} was not found , try again",
+                username, pattern,
+            );
+            Self::send_message(message, &text, api).await?;
+            return Err(BotError::DbError(BotDbError::CityNotFoundError));
+        }
+
+        let mut i = 1;
+        let mut text: String = format!(
+            "Hi {}, i found these cities put a number to select one\n\n",
+            username
+        );
+        for row in vec {
+            let name: String = row.get("name");
+            let country: String = row.get("country");
+            let state: String = row.get("state");
+            if state.is_empty() {
+                writeln!(&mut text, "{}. {},{}", i, name, country)?;
+            } else {
+                writeln!(&mut text, "{}. {},{},{}", i, name, country, state)?;
+            }
+            i += 1;
+        }
+        Self::send_message(message, &text, api).await?;
+
+        Ok(())
+    }
+
+    async fn pattern_response(
+        db_controller: &DbController,
+        message: &Message,
+        api: &AsyncApi,
+    ) -> Result<(), BotError> {
+        let (chat_id, user_id, username) = Self::get_info_from_message(message);
+
+        let number: usize = message.text.as_ref().unwrap().parse::<usize>().unwrap();
+
+        let selected = db_controller.get_client_selected(&chat_id, user_id).await?;
+
+        let (name, country, state) = match db_controller.get_city_row(&selected, number).await {
+            Ok((n, c, s)) => (n, c, s),
+            Err(_) => (String::new(), String::new(), String::new()),
+        };
+        let n: usize = match state.as_str() {
+            "" => 2,
+            _ => 3,
+        };
+
+        match db_controller
+            .get_client_before_state(&chat_id, user_id)
+            .await?
+        {
+            ClientState::Initial => Ok(Self::get_weather(
```

Comment:
get_weather accepts a lot of positional arguments. what do you think about creating a struct to pass it as an argument

Reply:
That was Conf struct